src/OpenTelemetry/Internal/CircularBuffer.cs

```diff
@@ -0,0 +1,135 @@
+﻿// <copyright file="CircularBuffer.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace OpenTelemetry.Internal
+{
+    /// <summary>
+    /// Lock free implementation of single-reader multi-writer circular buffer.
+    /// </summary>
+    /// <typeparam name="T">The type of the underlying value.</typeparam>
+    internal class CircularBuffer<T>
+        where T : class
+    {
+        private readonly int capacity;
+        private readonly T[] trait;
+        private long head = 0;
+        private long tail = 0;
+
+        public CircularBuffer(int capacity)
```

Comment:
Restrict capacity to be power of 2, so the modulo operation `%` could replaced with a simple bitwise and?

Reply:
How much (X%) perf improvement would you expect from modulo?

Reply:
`div` could take tens times of CPU cycles than bitwise arithmetic instruction, but there is also trade off taking more memory (for `ceiling(2**size)`). This could be a low priority optimization and needs data to prove.